homeworks/solutions/krasnor/5-remesh/src/halfedge.js

```diff
@@ -0,0 +1,821 @@
+
+    for(let i = 0; (curr_face_cnt > reduce_face_goal) && (curr_face_cnt > 0); i++){
+      let faces_removed = 0;
+      let p = heap.pop();
+      let e = p.pair; // Edge
+      let m = p.mergePoint;
+      if(e.idx === -1)
+        continue; // outdated edge, ignore
+
+      let real_face_cnt = 0;
+      let deleted_face_cnt = 0;
+      this.faces.forEach((v_f, index) => {
+          if(v_f.idx != -1){
+            real_face_cnt++;
+          }else{
+            deleted_face_cnt++;
+          }
+        }
+      );
+
+      console.log("in collapse: %s - current cnt faces %s || real face count: %s, deleted: %s", i, curr_face_cnt, real_face_cnt, deleted_face_cnt);
+      let curr_hf = e.halfedge;
+      let curr_hf_twin = curr_hf.twin;
+      let curr_p1 = curr_hf.vertex;
+      let curr_p2 = curr_hf_twin.vertex;
+
+      // create new vertex along edge
+      let new_vertex = new Vertex();
+      new_vertex.idx = nxt_idx_vertices++;
+      new_vertex.position = (new Vector()).add(m);
+      new_vertex.halfedge = curr_hf.prev.twin;
+
+      if(new_vertex.halfedge.idx === -1){
+        console.log("new_vertex.halfedge.idx is invalid");
+      }
+
+      // connect old halfedges to new vertex
+      curr_p1.halfedges(
+          (hf, i) => {
+            // console.log("overwriting hf: %s, point: %s with point: %s" , hf.idx, hf.vertex.idx, new_vertex.idx)
+            hf.vertex = new_vertex;
+          }
+      );
+      curr_p2.halfedges(
+          (hf, i) => {
+            // console.log("overwriting hf: %s, point: %s with point: %s" , hf.idx, hf.vertex.idx, new_vertex.idx)
+            hf.vertex = new_vertex;
+          }
+      );
+
+      // merge faces and their edges
+      [curr_hf, curr_hf_twin].forEach(
+          i_hf => {
+          let f = i_hf.face;
+          if(f != null){
+            //     c
+            //    / ^
+            // L / f \ R
+            //  v     \
+            //  a----->b
+            //   i_hf
+
+            let currFaceIdx = f.idx;
+            let l = i_hf.prev;
+            let r = i_hf.next;
+            let l_e = l.edge;
+            let r_e = r.edge;
+            let l_t = l.twin;
+            let r_t = r.twin;
+            let new_edge = new Edge();
+            new_edge.idx = nxt_idx_edges++;
+            new_edge.halfedge = l_t;
+
+
+            // update errors
+            let c = l.vertex;
+            c.old_id = c.idx;
+            c.halfedge = r_t;
+
+            l_t.twin = r_t;
+            r_t.twin = l_t;
+            l_t.edge.idx = -1;
+            r_t.edge.idx = -1;
+            l_t.edge = new_edge;
+            r_t.edge = new_edge;
+
+            // mark as obsolete
+            l.idx = -1;
+            r.idx = -1;
+            l_e.idx = -1;
+            r_e.idx = -1;
+            f.idx = -1;
+
+            this.edges.push(new_edge);
+            faces_removed++;
+          }
+        }
+      );
```

Comment:
Face by face removal. Indeed works but seems very hard to revisit. Do you still have this code in mind?

Reply:
Sorry, I should have provided more documentation.
A better name would probably be "dissolve Face".
Here the faces of an to be collapsed edge will be dissolved.

This is done based on the faces of `curr_hf` and `curr_hf.twin`.

In each loop the face, all inner halfedges and edges are marked as obsolete (idx = -1) and
`l_t` and `r_t` are merged to form a new edge.

```
// Diagram for first loop iteration - curr_hf:
// [1]      C                 C       
//        ^/^\               ^| 
//       //  \\              || 
//      //    \\             || 
//  l_t//L    R\\r_t  =>  l_t||r_t
//    //  (f)   \\           || 
//   /v          \v          |v 
//   P1-curr_hf--->P2     new_vert
//     <-c.hf.twin-   

// [0] new_vert is the midpopint of P1 and P2
// [1] P1 and P2 are actually already set to be new_vert through the previus steps
```